File: libs/go/waterfall.go
// Waterfall chart renderer — running-total floating bars with connectors.
// Byte-identical SVG output with the Python renderer.
//
// Proprietary. Copyright (c) 2026 Dharmik Shingala. All rights reserved.
package stonecharts

import (
	"fmt"
	"strings"
)

const waterfallPad = 0.2

func renderWaterfallSVG(spec *ChartSpec) string {
	s2 := *spec
	s2.Series = make([]Series, len(spec.Series))
	copy(s2.Series, spec.Series)

	sumIdx := make(map[int]bool)
	for _, v := range s2.SumIndices {
		sumIdx[v] = true
	}
	isumIdx := make(map[int]bool)
	for _, v := range s2.IntermediateSumIndices {
		isumIdx[v] = true
	}

	// Compute y-domain from running totals.
	allVals := []float64{0.0}
	for _, s := range s2.Series {
		running := 0.0
		for i, delta := range s.Data {
			if sumIdx[i] || isumIdx[i] {
				allVals = append(allVals, 0.0, running)
			} else {
				allVals = append(allVals, running)
				running += delta
				allVals = append(allVals, running)
			}
		}
	}
	if s2.YAxis.Min == nil {
		mn := allVals[0]
		for _, v := range allVals[1:] {
			if v < mn {
				mn = v
			}
		}
		s2.YAxis.Min = &mn
	}
	if s2.YAxis.Max == nil {
		mx := allVals[0]
		for _, v := range allVals[1:] {
			if v > mx {
				mx = v
			}
		}
		s2.YAxis.Max = &mx
	}

	return renderCartesian(&s2, "Waterfall", "band", waterfallMarks, true, "vertical")
}

type waterfallBar struct {
	start, end float64
	kind       string // "increase", "decrease", "total"
	totalAfter float64
}

func waterfallMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}
	spec := f.spec

	sumIdx := make(map[int]bool)
	for _, v := range spec.SumIndices {
		sumIdx[v] = true
	}
	isumIdx := make(map[int]bool)
	for _, v := range spec.IntermediateSumIndices {
		isumIdx[v] = true
	}

	upColor := spec.UpColor
	if upColor == "" {
		upColor = "#3f9b6a"
	}
	downColor := spec.DownColor
	if downColor == "" {
		downColor = "#d65f5f"
	}
	totalColor := spec.TotalColor
	if totalColor == "" {
		totalColor = "#4b6cb7"
	}

	connEnabled := true
	connDash := "4 3"
	connColor := f.theme.GridColor
	if spec.Connector != nil {
		if spec.Connector.Enabled != nil && !*spec.Connector.Enabled {
			connEnabled = false
		}
		if spec.Connector.DashStyle == "dotted" {
			connDash = "2 3"
		} else if spec.Connector.DashStyle == "solid" {
			connDash = ""
		}
	}

	band := f.bandWidth()
	groupW := band * (1 - waterfallPad)
	k := 1
	if f.spec.groupingOn() {
		k = len(f.spec.Series)
		if k < 1 {
			k = 1
		}
	}
	barW := groupW / float64(k)

	for si, s := range spec.Series {
		n := len(s.Data)
		if n > f.n {
			n = f.n
		}

		p.WriteString(fmt.Sprintf(`<g class="sc-series" data-series="%d">`, si))

		// Running-total transform.
		bars := make([]waterfallBar, n)
		running := 0.0
		for i := 0; i < n; i++ {
			delta := s.Data[i]
			if sumIdx[i] || isumIdx[i] {
				bars[i] = waterfallBar{0.0, running, "total", running}
			} else {
				start := running
				end := running + delta
				running = end
				kind := "increase"
				if delta < 0 {
					kind = "decrease"
				}
				bars[i] = waterfallBar{start, end, kind, running}
			}
		}

		siOff := 0
		if f.spec.groupingOn() {
			siOff = si
		}

		// Connectors first.
		if connEnabled {
			for i := 0; i < len(bars)-1; i++ {
				level := bars[i].totalAfter
				x1 := f.xpix(float64(i)) - groupW/2 + barW*float64(siOff) + barW
				x2 := f.xpix(float64(i+1)) - groupW/2 + barW*float64(siOff)
				y := f.ypix(level)
				dashAttr := ""
				if connDash != "" {
					dashAttr = ` stroke-dasharray="` + connDash + `"`
				}
				p.WriteString(fmt.Sprintf(
					`<line class="sc-connector" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s" stroke-width="1"%s/>`,
					f1(x1), f1(y), f1(x2), f1(y), connColor, dashAttr))
			}
		}

		// Bars.
		for i := 0; i < len(bars); i++ {
			b := bars[i]
			xlabel := fmt.Sprintf("%d", i)
			if i < len(f.cats) {
				xlabel = f.cats[i]
			}

			left := f.xpix(float64(i)) - groupW/2 + barW*float64(siOff)
			cx := left + barW/2

			yStart := f.ypix(b.start)
			yEnd := f.ypix(b.end)
			yTop := yEnd
			if yStart < yEnd {
				yTop = yStart
			}
			barH := yStart - yEnd
			if barH < 0 {
				barH = -barH
			}
			if barH < 1.0 {
				barH = 1.0
			}

			fill := upColor
			if b.kind == "decrease" {
				fill = downColor
			} else if b.kind == "total" {
				fill = totalColor
			}

			displayVal := b.totalAfter
			if b.kind != "total" {
				displayVal = s.Data[i]
			}

			p.WriteString(fmt.Sprintf(
				`<rect class="sc-bar sc-point" data-series="%d"`+
					` data-series-name="%s" data-x="%s"`+
					` data-y="%s"`+
					` data-kind="%s" data-total="%s"`+
					` data-color="%s" data-r="3.5" data-r-hover="6"`+
					` cx="%s" cy="%s"`+
					` x="%s" y="%s"`+
					` width="%s" height="%s"`+
					` fill="%s"/>`,
				si, esc(s.Name), esc(xlabel),
				esc(fmtNum(displayVal)),
				b.kind, esc(fmtNum(b.totalAfter)),
				fill,
				f1(cx), f1(yTop),
				f1(left), f1(yTop),
				f1(barW), f1(barH),
				fill))
		}

		p.WriteString(`</g>`)
	}
}

File: libs/python/stonecharts/charts/waterfall.py
"""Waterfall chart renderer: ChartSpec -> SVG string.

Floating bars stepping through signed deltas with a running total.
Shared Cartesian chrome comes from _cartesian.py; this module draws only the
waterfall marks: connector lines + one floating rect per stage.
"""

from __future__ import annotations

import copy

from ..spec import ChartSpec
from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, render_cartesian

PAD = 0.2


def render_svg(spec: ChartSpec) -> str:
    mod = copy.copy(spec)
    mod.y_axis = copy.copy(spec.y_axis)

    sum_idx = set(getattr(mod, "sum_indices", None) or [])
    isum_idx = set(getattr(mod, "intermediate_sum_indices", None) or [])

    # Running-total transform to compute y-domain.
    all_vals = [0.0]
    for s in mod.series:
        running = 0.0
        for i, delta in enumerate(s.data):
            if i in sum_idx or i in isum_idx:
                all_vals.append(0.0)
                all_vals.append(running)
            else:
                all_vals.append(running)
                running += delta
                all_vals.append(running)
    if mod.y_axis.min is None:
        mod.y_axis.min = min(all_vals)
    if mod.y_axis.max is None:
        mod.y_axis.max = max(all_vals)

    return render_cartesian(mod, "Waterfall", "band", _waterfall_marks)


def _waterfall_marks(fr: CartesianFrame, p: list[str]) -> None:
    if fr.n <= 0:
        return

    spec = fr.spec
    sum_idx = set(getattr(spec, "sum_indices", None) or [])
    isum_idx = set(getattr(spec, "intermediate_sum_indices", None) or [])
    up_color = getattr(spec, "up_color", "#3f9b6a")
    down_color = getattr(spec, "down_color", "#d65f5f")
    total_color = getattr(spec, "total_color", "#4b6cb7")

    connector_cfg = getattr(spec, "connector", None)
    conn_enabled = True
    conn_dash = "4 3"  # dashed default
    conn_color = fr.theme.grid_color
    if connector_cfg is not None:
        if hasattr(connector_cfg, "enabled"):
            conn_enabled = connector_cfg.enabled
        if hasattr(connector_cfg, "dash_style"):
            ds = connector_cfg.dash_style
            if ds == "dotted":
                conn_dash = "2 3"
            elif ds == "solid":
                conn_dash = ""

    band = fr.band_width()
    group_w = band * (1 - PAD)
    k = 1 if not fr.spec.grouping else max(len(fr.spec.series), 1)
    bar_w = group_w / k

    for si, s in enumerate(spec.series):
        n = min(len(s.data), fr.n)
        p.append(f'<g class="sc-series" data-series="{si}">')

        # Running-total transform
        bars = []  # list of (start, end, kind, running_after)
        running = 0.0
        for i in range(n):
            delta = s.data[i]
            if i in sum_idx or i in isum_idx:
                start = 0.0
                end = running
                kind = "total"
                bars.append((start, end, kind, running))
            else:
                start = running
                end = running + delta
                running = end
                kind = "increase" if delta >= 0 else "decrease"
                bars.append((start, end, kind, running))

        # Emit connectors first (so bars paint over line ends)
        if conn_enabled:
            for i in range(len(bars) - 1):
                _start, _end, _kind, level = bars[i]
                x1 = fr.xpix(i) - group_w / 2 + bar_w * (si if fr.spec.grouping else 0) + bar_w
                x2 = fr.xpix(i + 1) - group_w / 2 + bar_w * (si if fr.spec.grouping else 0)
                y = fr.ypix(level)
                dash_attr = f' stroke-dasharray="{conn_dash}"' if conn_dash else ""
                p.append(
                    f'<line class="sc-connector" x1="{x1:.1f}" y1="{y:.1f}"'
                    f' x2="{x2:.1f}" y2="{y:.1f}"'
                    f' stroke="{conn_color}" stroke-width="1"{dash_attr}/>'
                )

        # Emit bars
        for i in range(len(bars)):
            start, end, kind, total_after = bars[i]
            xlabel = fr.cats[i] if i < len(fr.cats) else str(i)

            left = fr.xpix(i) - group_w / 2 + bar_w * (si if fr.spec.grouping else 0)
            cx = left + bar_w / 2
            y_top = fr.ypix(max(start, end))
            y_bot = fr.ypix(min(start, end))
            bar_h = max(abs(y_bot - y_top), 1.0)

            if kind == "increase":
                fill = up_color
            elif kind == "decrease":
                fill = down_color
            else:
                fill = total_color

            display_val = total_after if kind == "total" else s.data[i]

            p.append(
                f'<rect class="sc-bar sc-point" data-series="{si}"'
                f' data-series-name="{esc(s.name)}" data-x="{esc(xlabel)}"'
                f' data-y="{esc(fmt_num(display_val))}"'
                f' data-kind="{kind}" data-total="{esc(fmt_num(total_after))}"'
                f' data-color="{fill}" data-r="3.5" data-r-hover="6"'
                f' cx="{cx:.1f}" cy="{y_top:.1f}"'
                f' x="{left:.1f}" y="{y_top:.1f}"'
                f' width="{bar_w:.1f}" height="{bar_h:.1f}"'
                f' fill="{fill}"/>'
            )

        p.append("</g>")
